src/main/scala/ai/privado/rulevalidator/YamlFileValidator.scala

```diff
@@ -0,0 +1,195 @@
+  private val THREATS     = Source.fromInputStream(getClass.getResourceAsStream(s"${SCHEMA_DIR_PATH}threats.json")).mkString
+  private val COLLECTIONS = Source.fromInputStream(getClass.getResourceAsStream(s"${SCHEMA_DIR_PATH}collections.json")).mkString
+
+  val mapper = new ObjectMapper(new YAMLFactory())
+
+  val factory: JsonSchemaFactory = JsonSchemaFactory
+    .builder(
+      JsonSchemaFactory
+        .getInstance(JSON_SCHEMA_VERSION)
+    ).objectMapper(mapper)
+    .build()
+
+  implicit def betterFilesExtended(file: File): BetterFileUtil = new BetterFileUtil(file)
+  implicit def jsonSchemaExtended(jsonSchema: JsonSchema): JsonSchemaExtension = new JsonSchemaExtension(jsonSchema)
+
+  /**
+    * Iterate recursively inside a directory path, filter for ".yaml" and ".yml" files
+    * Identify corresponding schema file for each file and validate.
+    * Filter all files to collect files with Validation Errors
+    * @return an Iterator on ai.privado.rulevalidator.ValidationFailure objects
+    */
+  def validateDirectory(dir: File): Iterator[ValidationFailure] = {
+
+    val validationErrors: Iterator[ValidationFailure] = dir
+      .listRecursively
+      .filter(
+        subDir =>
+          subDir
+            .extension(toLowerCase=true)
+            .toString
+            .contains(".yaml")
+            ||
+            subDir
+              .extension(toLowerCase=true)
+              .toString
+              .contains(".yml")
+      )
+      .map(
+        ruleFile => {
+          val yamlAsJson = mapper
+            .readTree(
+              ruleFile.contentAsString()
+            )
+          val schemaFile = matchSchemaFile(ruleFile, yamlAsJson, CommandConstants.VALIDATE)
+          if (schemaFile != null)
+            ValidationFailure(ruleFile.validateJsonFile(schemaFile, yamlAsJson), ruleFile)
+          else null
+        }
+      ).filter(vf => vf != null)
+
+    validationErrors
+      .filter(
+        ve =>
+          !ve.validationMessages.isEmpty
+      )
+  }
+
+  /**
+    * Validate a single rule file.
+    * Identify corresponding schema file for the input file and validate.
+    * @param ruleFile better.files.File object of the rule file to be validated
+    * @param configDirectory better.files.Files config location
+    * @return Boolean stating whether the rule file is valid
+    */
+  def isValidRuleFile(ruleFile: File, configDirectory: File): Boolean = {
+    if (!ruleFile.pathAsString.contains(s"${configDirectory.pathAsString}/$RULES_FOLDER_IN_CONFIG_DIR")) {
+      return false
+    }
+    val yamlAsJson = mapper
+      .readTree(
+        ruleFile.contentAsString()
+      )
+    val schemaFileContent = matchSchemaFile(ruleFile, yamlAsJson)
+    if (schemaFileContent == null)
+      false
+    else {
+      val validationMessages = ruleFile.validateJsonFile(schemaFileContent, yamlAsJson).asScala
+      if (validationMessages.nonEmpty) {
```

Comment:
The core logic of validation is similar in `isValidRuleFile` and `validateDirectory` if possible can we have the core logic in a single function - ex 
```
val yamlAsJson = mapper
      .readTree(
        ruleFile.contentAsString()
      )
    val schemaFileContent = matchSchemaFile(ruleFile, yamlAsJson)
    if (schemaFileContent == null)
      false
    else {
      val validationMessages = ruleFile.validateJsonFile(schemaFileContent, yamlAsJson).asScala```